lib/srv/server/installer/autodiscover.go

```diff
@@ -0,0 +1,600 @@
+			"error", err,
+			"version", api.Version)
+		return api.Version
+	}
+
+	autoUpgradesVersionGetter := version.NewBasicHTTPVersionGetter(upgradeURL)
+	targetVersion, err := autoUpgradesVersionGetter.GetVersion(ctx)
+	if err != nil {
+		asi.Logger.WarnContext(ctx, "Failed to query target version, using api version",
+			"error", err,
+			"version", api.Version)
+		return api.Version
+	}
+	asi.Logger.InfoContext(ctx, "Found target version",
+		"channel_url", asi.autoUpgradesChannelURL,
+		"version", targetVersion)
+
+	return strings.TrimSpace(strings.TrimPrefix(targetVersion, "v"))
+}
+
+func fetchNodeAutoDiscoverLabels(ctx context.Context, imdsClient imds.Client) (map[string]string, error) {
+	nodeLabels := make(map[string]string)
+
+	switch imdsClient.GetType() {
+	case types.InstanceMetadataTypeAzure:
+		azureIMDSClient, ok := imdsClient.(*azureimds.InstanceMetadataClient)
+		if !ok {
+			return nil, trace.BadParameter("failed to obtain azure imds client")
+		}
+
+		instanceInfo, err := azureIMDSClient.GetInstanceInfo(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		nodeLabels[types.SubscriptionIDLabel] = instanceInfo.SubscriptionID
+		nodeLabels[types.VMIDLabel] = instanceInfo.VMID
+		nodeLabels[types.RegionLabel] = instanceInfo.Location
+		nodeLabels[types.ResourceGroupLabel] = instanceInfo.ResourceGroupName
+
+	case types.InstanceMetadataTypeEC2:
+		awsIMDSClient, ok := imdsClient.(*awsimds.InstanceMetadataClient)
+		if !ok {
+			return nil, trace.BadParameter("failed to obtain ec2 imds client")
+		}
+		accountID, err := awsIMDSClient.GetAccountID(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		instanceID, err := awsIMDSClient.GetID(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		nodeLabels[types.AWSInstanceIDLabel] = instanceID
+		nodeLabels[types.AWSAccountIDLabel] = accountID
+
+	case types.InstanceMetadataTypeGCP:
+		gcpIMDSClient, ok := imdsClient.(*gcpimds.InstanceMetadataClient)
+		if !ok {
+			return nil, trace.BadParameter("failed to obtain gcp imds client")
+		}
+
+		name, err := gcpIMDSClient.GetName(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		zone, err := gcpIMDSClient.GetZone(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		projectID, err := gcpIMDSClient.GetProjectID(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		nodeLabels[types.NameLabel] = name
+		nodeLabels[types.ZoneLabel] = zone
+		nodeLabels[types.ProjectIDLabel] = projectID
+
+	default:
+		return nil, trace.BadParameter("Unsupported cloud provider: %v", imdsClient.GetType())
+	}
+
+	return nodeLabels, nil
+}
+
+// buildAbsoluteFilePath creates the absolute file path
+func (asi *AutodiscoverNodeInstaller) buildAbsoluteFilePath(filepath string) string {
+	return path.Join(asi.fsRootPrefix, filepath)
+}
+
+type linuxDistroInfo struct {
+	*linux.OSRelease
+}
+
+func (l *linuxDistroInfo) packageManagerKind() packageManagerKind {
```

Comment:
Nit: What do you think about moving this (along with all related logic/structs/constants) somewhere to utils package? Feels like this could be useful outside of just installer in future e.g. for the next-gen auto-updates work.

Reply:
Moved things to `lib/utils/packagemanager/` and removed some not-required types/methods.